ast/src/lib.rs

```diff
@@ -77,32 +77,40 @@ where
     write!(f, "{}", indent(s, indentation))
 }
 
-fn write_items<S, I, W>(f: &mut W, items: I) -> Result
+fn write_items<S, I, W>(f: &mut W, items: I, line_end: Option<&str>) -> Result
 where
     S: Display,
     I: IntoIterator<Item = S>,
     W: Write,
 {
-    write_items_indented_by(f, items, 0)
+    write_items_indented_by(f, items, 0, line_end)
 }
 
-fn write_items_indented<S, I, W>(f: &mut W, items: I) -> Result
+fn write_items_indented<S, I, W>(f: &mut W, items: I, line_end: Option<&str>) -> Result
 where
     S: Display,
     I: IntoIterator<Item = S>,
     W: Write,
 {
-    write_items_indented_by(f, items, 1)
+    write_items_indented_by(f, items, 1, line_end)
 }
 
-fn write_items_indented_by<S, I, W>(f: &mut W, items: I, by: usize) -> Result
+fn write_items_indented_by<S, I, W>(
+    f: &mut W,
+    items: I,
+    by: usize,
+    line_end: Option<&str>,
+) -> Result
 where
     S: Display,
     I: IntoIterator<Item = S>,
     W: Write,
 {
     for item in items.into_iter() {
         write_indented_by(f, item, by)?;
```

Comment:
I never really liked this indentation mechanism. It forces every multi-line Display implementation to use the same mechanism and I have the impression that requiring this special line_end parameter is a symptom of that.
Also it does not work for nested structures.

Could you check if you could use the indentation mechanism used by Expression (i.e. just format to a string and then call `indent` on that string)?

Reply:
ah wait, this uses that mechanism under the hood. Ok, I think I need to take a closer look.